pkg/chunked/storage_linux.go

```diff
@@ -413,18 +449,46 @@ func createFileFromZstdStream(dest string, dirfd int, reader io.Reader, mode os.
 		}
 	}()
 
-	z, err := zstd.NewReader(reader)
-	if err != nil {
-		return err
-	}
-	defer z.Close()
-
 	digester := digest.Canonical.Digester()
 	checksum := digester.Hash()
-	_, err = z.WriteTo(io.MultiWriter(file, checksum))
-	if err != nil {
-		return err
+	to := io.MultiWriter(file, checksum)
+
+	switch c.fileType {
+	case fileTypeZstdChunked:
+		z, err := zstd.NewReader(reader)
+		if err != nil {
+			return err
+		}
+		defer z.Close()
+
+		_, err = z.WriteTo(to)
+		if err != nil {
+			return err
+		}
+	case fileTypeEstargz:
+		if c.gzipReader == nil {
+			r, err := pgzip.NewReader(reader)
+			if err != nil {
+				return err
+			}
+			c.gzipReader = r
+		} else {
+			if err := c.gzipReader.Reset(reader); err != nil {
+				return err
+			}
+		}
+		defer c.gzipReader.Close()
+
+		if _, err := io.Copy(to, io.LimitReader(c.gzipReader, metadata.Size)); err != nil {
+			return err
+		}
+		if _, err := io.Copy(ioutil.Discard, c.gzipReader); err != nil {
+			return err
+		}
```

Comment:
Could we do `io.Copy(ioutil.Discard, reader)` instead of `c.gzipReader` so that we can avoid any parse errors that occurred from this discarding part?